### ISSUE TITLE
Two Approaches to Enhanced Formulae Language

### ISSUE THREAD

**Issue by Nosferican** [JuliaStats/StatsModels.jl#29]
Currently `@formula` macro generates the `Formula` struct which has fields: `lhs` and `rhs`.  For categorical variables (`PooledDataArray`), `ModelFrame` can use a `Dict{Symbol, ContrastsMatrix}` to prepare the object for `ModelMatrix` to generate the desired `Matrix` and keep track of relevant data (e.g., `coefnames`).

### First Approach
The feature request is to allow for additional operators other than the currently implemented `+`, `*`, `&`. For categorical variables, codings are enabled through
```
mf = ModelFrame(@formula(y ~ x), df, contrasts = Dict(:x => EffectsCoding()))
```
and the following codings are available
- DummyCoding (default)
- EffectsCoding
- HelmertCoding
- ContrastsCoding

A welcomed addition would be to allow for codings for numerical variables (`DataArrays.DataArray`). For example,
- IdentityCoding (default)
- Log
- Poly(d::Integer)
```
fm = DataFrames.@formula(y ~ x1 + x2)
mf = DataFrames.ModelFrame(fm, df, contrasts = Dict((:x1, LogCoding()), (:x2, PolyCoding(2)))
```

### Second Approach
Inline specification can be can work as:
- Stata's `reg y i.x` for dummy coding
- R's `lm(y ~ log(x), df)` for common transformation
and implement these as we currently handle `+`, `*` and `&` for interactions.
```
@formula(y ~ log(x1) + poly(x2, 2) + i.x3 + EffectsCoding(x4))
```
see [StreamModels](https://github.com/kleinschmidt/StreamModels.jl).

Additional thought:
- The polynomial transformation can be pretty useful especially when building models where keeping track of derived features is essential. For example, for accurately computing the marginal effects of a logistic regression. Other functions that might use the information are multicollinearity diagnostics or lasso / net-elastic estimators in the group-versions.
- While the second approach seems more convenient, the first one tends to be more robust. See for example  #27 where having more information on the coding for each variable is preferred (e.g., reference level). A mixture could give a best of both solution by having `Formula` or `ModelFrame` additional fields for holding this information.

**Comment by nalimilan** [comment]
Indeed it can be useful to save information about the transformation, for example in the case of splines, where the position of the knots is necessary to interpret the coefficients. Having a `splines(x, 3)` term in the formula would not be enough.

However, it shouldn't be called "coding", as it's quite different from the contrasts system used for categorical variables. I'd rather call it "transformation" or "derived".

**Comment by kleinschmidt** [comment]
Right. My plan (in StreamModels) is to save all that information in `Term` types (e.g., a `CategoricalTerm` for contrast coded variables, a `SpilneTerm` for splines).
